File: src/src_data/fmg/data/view/draw/ColorText.java
package fmg.data.view.draw;

import java.beans.PropertyChangeListener;
import java.util.Arrays;

import fmg.common.Color;
import fmg.common.notyfier.INotifyPropertyChanged;
import fmg.common.notyfier.NotifyPropertyChanged;
import fmg.core.types.EClose;
import fmg.core.types.EOpen;

public class ColorText implements INotifyPropertyChanged {

   private Color[] colorOpen;
   private Color[] colorClose;
   protected NotifyPropertyChanged _notifier = new NotifyPropertyChanged(this);

   public ColorText() {
      colorOpen = new Color[EOpen.values().length];
      colorClose = new Color[EClose.values().length];

      for (EOpen eOpen: EOpen.values())
         switch (eOpen) {
         case _Nil : colorOpen[eOpen.ordinal()] = Color.Black ; break;
         case _1   : colorOpen[eOpen.ordinal()] = Color.Navy  ; break;
         case _2   : colorOpen[eOpen.ordinal()] = Color.Green ; break;
         case _3   : colorOpen[eOpen.ordinal()] = Color.Red   ; break;
         case _4   : colorOpen[eOpen.ordinal()] = Color.Maroon; break;
         case _5   : colorOpen[eOpen.ordinal()] = Color.Blue  ; break;
         case _6   : colorOpen[eOpen.ordinal()] = Color.Black ; break;
         case _7   : colorOpen[eOpen.ordinal()] = Color.Olive ; break;
         case _8   : colorOpen[eOpen.ordinal()] = Color.Aqua  ; break;
         case _9   : colorOpen[eOpen.ordinal()] = Color.Navy  ; break;
         case _10  : colorOpen[eOpen.ordinal()] = Color.Green ; break;
         case _11  : colorOpen[eOpen.ordinal()] = Color.Red   ; break;
         case _12  : colorOpen[eOpen.ordinal()] = Color.Maroon; break;
         case _13  : colorOpen[eOpen.ordinal()] = Color.Navy  ; break;
         case _14  : colorOpen[eOpen.ordinal()] = Color.Green ; break;
         case _15  : colorOpen[eOpen.ordinal()] = Color.Red   ; break;
         case _16  : colorOpen[eOpen.ordinal()] = Color.Maroon; break;
         case _17  : colorOpen[eOpen.ordinal()] = Color.Blue  ; break;
         case _18  : colorOpen[eOpen.ordinal()] = Color.Black ; break;
         case _19  : colorOpen[eOpen.ordinal()] = Color.Olive ; break;
         case _20  : colorOpen[eOpen.ordinal()] = Color.Aqua  ; break;
         case _21  : colorOpen[eOpen.ordinal()] = Color.Navy  ; break;
         case _Mine: colorOpen[eOpen.ordinal()] = Color.Black ; break;
         default: throw new RuntimeException("add EOpen value");
         }

      for (EClose eClose: EClose.values())
         switch (eClose) {
         case _Unknown: colorClose[eClose.ordinal()] = Color.Teal ; break;
         case _Clear  : colorClose[eClose.ordinal()] = Color.Black; break;
         case _Flag   : colorClose[eClose.ordinal()] = Color.Red  ; break;
         default: throw new RuntimeException("add EClose value");
         }
   }

   public static final String PROPERTY_COLOR_OPEN    = "ColorOpen";
   public static final String PROPERTY_COLOR_OPEN_N_ = "ColorOpen.#";
   public static final String PROPERTY_COLOR_CLOSE   = "ColorClose";
   public static final String PROPERTY_COLOR_CLOSE_N = "ColorClose.#";

   public Color[] getColorOpen() {
      return colorOpen;
   }
   public Color getColorOpen(int i) {
      return colorOpen[i];
   }
   public void setColorOpen(Color[] colorOpen) {
      Color[] old = this.colorOpen;
      if (!Arrays.equals(old, colorOpen)) {
         this.colorOpen = colorOpen;
         _notifier.onPropertyChanged(old, colorOpen, PROPERTY_COLOR_OPEN);
      }
   }
   public void setColorOpen(int i, Color colorOpen) {
      Color old = this.colorOpen[i];
      if (!old.equals(colorOpen)) {
         this.colorOpen[i] = colorOpen;
         _notifier.onPropertyChanged(old, colorOpen, PROPERTY_COLOR_OPEN_N_ + i);
      }
   }

   public Color[] getColorClose() {
      return colorClose;
   }
   public Color getColorClose(int i) {
      return colorClose[i];
   }
   public void setColorClose(Color[] colorClose) {
      Color[] old = this.colorClose;
      if (!Arrays.equals(old, colorClose)) {
         this.colorClose = colorClose;
         _notifier.onPropertyChanged(old, colorClose, PROPERTY_COLOR_CLOSE);
      }
   }
   public void setColorClose(int i, Color colorClose) {
      Color old = this.colorClose[i];
      if (!old.equals(colorClose)) {
         this.colorClose[i] = colorClose;
         _notifier.onPropertyChanged(old, colorClose, PROPERTY_COLOR_CLOSE_N + i);
      }
   }

   @Override
   public int hashCode() {
      int result = 31 + Arrays.hashCode(colorClose);
      return 31 * result + Arrays.hashCode(colorOpen);
   }

   @Override
   public boolean equals(Object obj) {
      if (this == obj) return true;
      if (!(obj instanceof ColorText)) return false;
      ColorText other = (ColorText) obj;
      return Arrays.equals(colorClose, other.colorClose) &&
             Arrays.equals(colorOpen, other.colorOpen);
   }

   @Override
   public void addListener(PropertyChangeListener listener) {
      _notifier.addListener(listener);
   }
   @Override
   public void removeListener(PropertyChangeListener listener) {
      _notifier.removeListener(listener);
   }

}

File: FastMines_core/src/fmg/core/types/draw/ColorText.cs
using Fmg.Common;
using Fmg.Common.Notifier;
using System;
using System.ComponentModel;

namespace Fmg.Core.Types.Draw {

    public class ColorText : INotifyPropertyChanged, IDisposable {

        private readonly Color[] _colorOpen;
        private readonly Color[] _colorClose;
        public event PropertyChangedEventHandler PropertyChanged {
            add    { _notifier.PropertyChanged += value;  }
            remove { _notifier.PropertyChanged -= value;  }
        }
        protected readonly NotifyPropertyChanged _notifier;

        public ColorText() {
            _notifier = new NotifyPropertyChanged(this);
            _colorOpen = new Color[EOpenEx.GetValues().Length];
            _colorClose = new Color[ECloseEx.GetValues().Length];

            foreach (var eOpen in EOpenEx.GetValues())
                switch (eOpen) {
                case EOpen._Nil : _colorOpen[eOpen.Ordinal()] = Color.Black;  break;
                case EOpen._1   : _colorOpen[eOpen.Ordinal()] = Color.Navy;   break;
                case EOpen._2   : _colorOpen[eOpen.Ordinal()] = Color.Green;  break;
                case EOpen._3   : _colorOpen[eOpen.Ordinal()] = Color.Red;    break;
                case EOpen._4   : _colorOpen[eOpen.Ordinal()] = Color.Maroon; break;
                case EOpen._5   : _colorOpen[eOpen.Ordinal()] = Color.Blue;   break;
                case EOpen._6   : _colorOpen[eOpen.Ordinal()] = Color.Black;  break;
                case EOpen._7   : _colorOpen[eOpen.Ordinal()] = Color.Olive;  break;
                case EOpen._8   : _colorOpen[eOpen.Ordinal()] = Color.Aqua;   break;
                case EOpen._9   : _colorOpen[eOpen.Ordinal()] = Color.Navy;   break;
                case EOpen._10  : _colorOpen[eOpen.Ordinal()] = Color.Green;  break;
                case EOpen._11  : _colorOpen[eOpen.Ordinal()] = Color.Red;    break;
                case EOpen._12  : _colorOpen[eOpen.Ordinal()] = Color.Maroon; break;
                case EOpen._13  : _colorOpen[eOpen.Ordinal()] = Color.Navy;   break;
                case EOpen._14  : _colorOpen[eOpen.Ordinal()] = Color.Green;  break;
                case EOpen._15  : _colorOpen[eOpen.Ordinal()] = Color.Red;    break;
                case EOpen._16  : _colorOpen[eOpen.Ordinal()] = Color.Maroon; break;
                case EOpen._17  : _colorOpen[eOpen.Ordinal()] = Color.Blue;   break;
                case EOpen._18  : _colorOpen[eOpen.Ordinal()] = Color.Black;  break;
                case EOpen._19  : _colorOpen[eOpen.Ordinal()] = Color.Olive;  break;
                case EOpen._20  : _colorOpen[eOpen.Ordinal()] = Color.Aqua;   break;
                case EOpen._21  : _colorOpen[eOpen.Ordinal()] = Color.Navy;   break;
                case EOpen._Mine: _colorOpen[eOpen.Ordinal()] = Color.Black;  break;
                default: throw new Exception("add EOpen value");
                }

            foreach (var eClose in ECloseEx.GetValues())
                switch (eClose) {
                case EClose._Unknown: _colorClose[eClose.Ordinal()] = Color.Teal;  break;
                case EClose._Clear  : _colorClose[eClose.Ordinal()] = Color.Black; break;
                case EClose._Flag   : _colorClose[eClose.Ordinal()] = Color.Red;   break;
                default: throw new Exception("add EClose value");
                }
        }

        public Color GetColorOpen(int i) {
            return _colorOpen[i];
        }

        public void SetColorOpen(int i, Color colorOpen) {
            _notifier.SetProperty(ref _colorOpen[i], colorOpen, "ColorOpen..#" + i);
        }

        public Color GetColorClose(int i) {
            return _colorClose[i];
        }

        public void SetColorClose(int i, Color colorClose) {
            _notifier.SetProperty(ref _colorClose[i], colorClose, "ColorClose.#" + i);
        }

        public void Dispose() {
            _notifier.Dispose();
            GC.SuppressFinalize(this);
        }

    }

}
